apps/builder/app/routes/rest.domain.ts

```diff
@@ -0,0 +1,78 @@
+import type { ActionArgs } from "@remix-run/node";
+import { zfd } from "zod-form-data";
+import { prisma } from "@webstudio-is/prisma-client";
+import { trpcClient } from "~/services/trpc.server";
+
+const schema = zfd.formData({
+  projectId: zfd.text(),
+  userId: zfd.text(),
+  customDomain: zfd.text(),
+  customCname: zfd.text(),
+});
+
+async function addDomain({
+  customDomain,
+  projectId,
+  userId,
+}: {
+  customDomain: string;
+  projectId: string;
+  userId: string;
+}) {
+  return await prisma.customDomain.create({
+    data: {
+      domain: customDomain,
+      projectId,
+      userId,
+    },
+  });
+}
+export const action = async ({ request }: ActionArgs) => {
+  const { customDomain, customCname, projectId, userId } = schema.parse(
+    await request.formData()
+  );
+
+  const customDomainFound = await prisma.customDomain.findFirst({
```

Comment:
It should be not a problem if a custom domain is found in our db.  Few users can enter the same domain, final routing depends on cname.